### PR TITLE
[InstCombine] Fix Failure to convert vector fp comparisons that can be represented as integers #82241

### LINKED ISSUE #82241: [InstCombine] Failure to convert vector fp comparisons that can be represented as integers
https://rust.godbolt.org/z/dfWjnToef

We will simplify a comparison of scalar floats that can be represented as integers to compare the integers directly, but we don't do this for vectors:

```ll
define i32 @cmpf(i32 %x) {
  %and = and i32 %x, 3
  %conv = sitofp i32 %and to float
  %cmp = fcmp oeq float %conv, 3.000000e+00
  %sext = sext i1 %cmp to i32
  ret i32 %sext
}

define <8 x i32> @vcmpf(<8 x i32> %x) {
  %and = and <8 x i32> %x, <i32 3, i32 3, i32 3, i32 3, i32 3, i32 3, i32 3, i32 3>
  %conv = sitofp <8 x i32> %and to <8 x float>
  %cmp = fcmp oeq <8 x float> %conv, <float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00>
  %sext = sext <8 x i1> %cmp to <8 x i32>
  ret <8 x i32> %sext
}
```
opt -O3
```ll
define i32 @cmpf(i32 %x) {
  %and = and i32 %x, 3
  %cmp = icmp eq i32 %and, 3
  %sext = sext i1 %cmp to i32
  ret i32 %sext
}

define <8 x i32> @vcmpf(<8 x i32> %x) {
  %and = and <8 x i32> %x, <i32 3, i32 3, i32 3, i32 3, i32 3, i32 3, i32 3, i32 3>
  %conv = sitofp <8 x i32> %and to <8 x float>
  %cmp = fcmp oeq <8 x float> %conv, <float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00, float 3.000000e+00>
  %sext = sext <8 x i1> %cmp to <8 x i32>
  ret <8 x i32> %sext
}
```

### PR DESCRIPTION
**Pull request by SahilPatidar**
Resolve #82241

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -7286,6 +7286,10 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
 Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
                                                     Instruction *LHSI,
                                                     Constant *RHSC) {
+  if (RHSC->getType()->isVectorTy()) {
+    if (Constant *CV = RHSC->getSplatValue(false))
+      RHSC = CV;
+  }
   if (!isa<ConstantFP>(RHSC)) return nullptr;
   const APFloat &RHS = cast<ConstantFP>(RHSC)->getValueAPF();
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7294]:
```suggestion
  const APFloat *RHS;
  if (!match(RHSC, m_APFloat(RHS)))
    return nullptr;
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -7298,6 +7302,16 @@ Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
...
+  Value *False, *True;
+  if (IntTy->isVectorTy()) {
+    ElementCount EC = cast<VectorType>(IntTy)->getElementCount();
+    True = Builder.CreateVectorSplat(EC, Builder.getTrue());
+    False = Builder.CreateVectorSplat(EC, Builder.getFalse());
+  } else {
+    True = Builder.getTrue();
+    False = Builder.getFalse();
+  }
+
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7314]:
```suggestion
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -7312,10 +7326,10 @@ Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
       RHSRoundInt.roundToIntegral(APFloat::rmNearestTiesToEven);
       if (RHS != RHSRoundInt) {
         if (P == FCmpInst::FCMP_OEQ || P == FCmpInst::FCMP_UEQ)
-          return replaceInstUsesWith(I, Builder.getFalse());
+          return replaceInstUsesWith(I, False);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7329]:
```suggestion
          return replaceInstUsesWith(I, ConstantInt::getFalse(I.getType()));
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 4]
```diff
@@ -7312,10 +7326,10 @@ Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
...
 
         assert(P == FCmpInst::FCMP_ONE || P == FCmpInst::FCMP_UNE);
-        return replaceInstUsesWith(I, Builder.getTrue());
+        return replaceInstUsesWith(I, True);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7332]:
```suggestion
        return replaceInstUsesWith(I, ConstantInt::getTrue(I.getType()));
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 5]
```diff
@@ -7511,6 +7525,13 @@ Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
     }
   }
 
+  if (IntTy->isVectorTy()) {
+    Value *RHSV = Builder.CreateVectorSplat(
+                     cast<VectorType>(IntTy)->getElementCount(),
+                     Builder.getInt(RHSInt));
+    return new ICmpInst(Pred, LHSI->getOperand(0), RHSV);
+  }
+
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7534]:
```suggestion
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 6]
```diff
@@ -7511,6 +7525,13 @@ Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
...
   // Lower this FP comparison into an appropriate integer version of the
   // comparison.
   return new ICmpInst(Pred, LHSI->getOperand(0), Builder.getInt(RHSInt));
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7537]:
```suggestion
  return new ICmpInst(Pred, LHSI->getOperand(0), ConstantInt::get(LHSI->getOperand(0)->getType(), RHSInt));
```

File: llvm/test/Transforms/InstCombine/cast-int-fcmp-eq-0.ll [hunk 7]
```diff
@@ -509,3 +509,61 @@ define i1 @i128_cast_cmp_oeq_int_inf_uitofp(i128 %i) {
   %cmp = fcmp oeq float %f, 0x7FF0000000000000
   ret i1 %cmp
 }
+
+define <2 x i1> @i32_vec_cast_cmp_oeq_vec_int_0_sitofp(<2 x i32> %i) {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/cast-int-fcmp-eq-0.ll:513]: Please pre-commit these tests.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 8]
```diff
@@ -7286,6 +7286,10 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
 Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
                                                     Instruction *LHSI,
                                                     Constant *RHSC) {
+  if (RHSC->getType()->isVectorTy()) {
+    if (Constant *CV = RHSC->getSplatValue(false))
+      RHSC = CV;
+  }
```
Comment (RKSimon) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7292]: No need to check isVectorTy - its safe to just do:
```cpp
if (Constant *CV = RHSC->getSplatValue(false))
  RHSC = CV;
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 9]
```diff
@@ -7298,6 +7302,16 @@ Instruction *InstCombinerImpl::foldFCmpIntToFPConst(FCmpInst &I,
   unsigned IntWidth = IntTy->getScalarSizeInBits();
   bool LHSUnsigned = isa<UIToFPInst>(LHSI);
 
+  Value *False, *True;
+  if (IntTy->isVectorTy()) {
+    ElementCount EC = cast<VectorType>(IntTy)->getElementCount();
+    True = Builder.CreateVectorSplat(EC, Builder.getTrue());
+    False = Builder.CreateVectorSplat(EC, Builder.getFalse());
```
Comment (RKSimon) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7309]: @nikic Is there a better way to do this - could we create Builder.getTrue(Type*) variants?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 8 above)
Comment (RKSimon) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7292]: @dtcxzyw m_APFloat proposal is better

**Review by RKSimon — CHANGES_REQUESTED** [review]

**Review by arsenm — APPROVED** [review]

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Review by RKSimon — APPROVED** [review]
LGTM - @SahilPatidar do you have commit access?

**Comment by arsenm** [comment]
> LGTM - @SahilPatidar do you have commit access?

Mouseover on the username usually shows if they're a member of llvm-committers

**Pull request merged**